### PR TITLE
Dynamic ISF update to model

### PR DESCRIPTION
**Pull request by tim2000s**
This push consists of the following changes:

DYnamic ISF Plugin uses combination of 8 hour weighted average, last full day and 7 day average TDDs.

TDD calculator adjusted to provide this information.
Model is the reworked 1800 rule using log values and insulin peak.

LGS Threshold moved in preferences to the DynamicISF prefs module and only available in the Dynamic ISF plugin.

### CODE REVIEW

File: app/src/main/assets/OpenAPSSMBDynamicISF/determine-basal.js [hunk 1]
```diff
@@ -501,9 +510,41 @@ var determine_basal = function determine_basal(glucose_status, currenttemp, iob_
...
+            lgsThreshold = (18 * lgsThreshold);
+            lgsThreshold = round(lgsThreshold, 2);
+            console.error("LGS Threshold converted to"+lgsThreshold+"; ");
+        }else{
+            lgsThreshold;
+        }
+        var threshold = min_bg - 0.5*(min_bg-40);
+        var oldThreshold = threshold;
+        if(lgsThreshold < 65 || lgsThreshold > 120) {
+            threshold;
```
Comment (MilosKozak) [inline app/src/main/assets/OpenAPSSMBDynamicISF/determine-basal.js:527]: It does nothing

File: app/src/main/assets/OpenAPSSMBDynamicISF/determine-basal.js [hunk 2]
```diff
@@ -501,9 +510,41 @@ var determine_basal = function determine_basal(glucose_status, currenttemp, iob_
...
+            }
+        else if( lgsThreshold < threshold){
+            threshold;
```
Comment (MilosKozak) [inline app/src/main/assets/OpenAPSSMBDynamicISF/determine-basal.js:530]: It does nothing

File: app/src/main/assets/OpenAPSSMBDynamicISF/determine-basal.js [hunk 3]
```diff
@@ -501,9 +510,41 @@ var determine_basal = function determine_basal(glucose_status, currenttemp, iob_
         return rT;
     }
 
-    // min_bg of 90 -> threshold of 65, 100 -> 70 110 -> 75, and 130 -> 85
-    var threshold = min_bg - 0.5*(min_bg-40);
+    // min_bg of 90 -> threshold of 65, 100 -> 70 110 -> 75, and 130 -> 85, or if specified by user, take that value
+        var lgsThreshold = profile.lgsThreshold;
+        console.error("Profile LGS Threshold set to"+lgsThreshold+"; ");
+        //if(lgsThreshold < 5.6)
+        if(lgsThreshold > 3.2 && lgsThreshold < 5.6){
```
Comment (MilosKozak) [inline app/src/main/assets/OpenAPSSMBDynamicISF/determine-basal.js:517]: All JS code code works in mgdl. I believe we must continue with this assumption and do conversion in kotlin

File: app/src/main/java/info/nightscout/androidaps/utils/stats/TddCalculator.kt [hunk 4]
```diff
@@ -161,6 +161,102 @@ class TddCalculator @Inject constructor(
...
+        tdd.totalAmount = tdd.bolusAmount + tdd.basalAmount
+        //}
+
+
+        aapsLogger.debug(LTag.CORE, tdd.toString())
+        return tdd
+    }
+
+
+    fun calculate8Gap():TotalDailyDose {
```
Comment (MilosKozak) [inline app/src/main/java/info/nightscout/androidaps/utils/stats/TddCalculator.kt:213]: this is copy-paste
replace with 1 function with params

File: app/src/main/assets/OpenAPSSMBDynamicISF/determine-basal.js (hunk 2 above)
Comment (tim2000s) [reply app/src/main/assets/OpenAPSSMBDynamicISF/determine-basal.js:530]: Yes. If the user LGS threshold is outside the limits that we allow, including the calculated threshold, we leave the calculated threshold. It's only when user defined value is higher than cacluated value, that we replace the calculated value.

File: app/src/main/assets/OpenAPSSMBDynamicISF/determine-basal.js (hunk 3 above)
Comment (tim2000s) [reply app/src/main/assets/OpenAPSSMBDynamicISF/determine-basal.js:517]: I've tried that. The conversion doesn't pick up using the Kotlin code. It only changes the display value and doesn't get passed to determine basal. 

Happy to have it in Kotlin, and would prefer it that way, but so far haven't been able to make it work.

File: app/src/main/java/info/nightscout/androidaps/utils/stats/TddCalculator.kt (hunk 4 above)
Comment (tim2000s) [reply app/src/main/java/info/nightscout/androidaps/utils/stats/TddCalculator.kt:213]: If you allow this to be used, then I can try and work that out, but right now that will take me some time, and I'd like to get the updated model into wider circulation.

**Comment by MilosKozak** [comment]
my cleanup https://github.com/nightscout/AndroidAPS/pull/1832/commits/2d6de1970b637222ce1951864748f520cf7cee72

**Pull request merged**